### PR TITLE
Add tests for some ICEs

### LINKED ISSUE #43623: ICE: Failed to unify obligation
The compiler ices with error "Failed to unify obligation" when I try to compile the following code:

```rust
trait StreamingIterator<'a> {
	type Item : 'a;
	fn next(&'a mut self) -> Option<Self::Item>;
	fn map<F>(self, f: F) -> StreamingIterMap<Self, F> 
	where Self: Sized
	{
		StreamingIterMap{iter: self, f: f}
	}
}

struct StreamingIterMap<I, F>
{
	iter: I,
	f: F
}

impl<'b, I, F, T> StreamingIterator<'b> for StreamingIterMap<I, F>
where
	I: for<'a:'b> StreamingIterator<'a>,
	F: for<'a:'b> Fn(<I as StreamingIterator<'a>>::Item) -> T,
	T: 'b,
{
	type Item = T;
	fn next(&mut self) -> Option<T> {
		self.iter.next().map(&self.f)
	}
}

struct StreamingCounter {
	index: usize
}

impl<'a> StreamingIterator<'a> for StreamingCounter  {
	type Item = &'a usize;
	fn next(&'a mut self) -> Option<&'a usize> {
		self.index += 1;
		Some(&self.index)
	}
}

fn main() {
	let mut svi = StreamingCounter{index: 0}
		.map(|u| Some(u));
	if let Some(ref i) = svi.next() {
		println!("{}", i);
	}
}
```

I get the following error:

```
error: internal compiler error: /checkout/src/librustc/traits/project.rs:1230: Failed to unify obligation `Obligation(predicate=ProjectionTy { trait_ref: <[closure@test.rs:43:8: 43:19] as std::ops::FnOnce<(<StreamingCounter as StreamingIterator<'a>>::Item,)>>, item_def_id: DefId { krate: CrateNum(2), node: DefIndex(895) => core/ce488d1::ops[0]::FnOnce[0]::Output[0] } },depth=0)` with poly_projection `Binder(ProjectionPredicate(ProjectionTy { trait_ref: <[closure@test.rs:43:8: 43:19] as std::ops::FnOnce<(&usize,)>>, item_def_id: DefId { krate: CrateNum(2), node: DefIndex(895) => core/ce488d1::ops[0]::FnOnce[0]::Output[0] } }, std::option::Option<_>))`: Sorts(ExpectedFound { expected: &usize, found: <StreamingCounter as StreamingIterator<'a>>::Item })
  --> test.rs:44:27
   |
44 | 	if let Some(ref i) = svi.next() {
   | 	                         ^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:428
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

## Meta

`rustc --version --verbose`:

```
rustc 1.19.0 (0ade33941 2017-07-17)
binary: rustc
commit-hash: 0ade339411587887bf01bcfa2e9ae4414c8900d4
commit-date: 2017-07-17
host: i686-unknown-linux-gnu
release: 1.19.0
LLVM version: 4.0
```

Backtrace:
```
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:365
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::span_bug
   7: rustc::session::opt_span_bug_fmt::{{closure}}
   8: rustc::session::opt_span_bug_fmt
   9: rustc::session::span_bug_fmt
  10: rustc::traits::project::confirm_param_env_candidate
  11: rustc::traits::project::confirm_callable_candidate
  12: rustc::traits::project::confirm_closure_candidate
  13: rustc::traits::project::confirm_candidate
  14: rustc::traits::project::opt_normalize_projection_type
  15: rustc::traits::project::poly_project_and_unify_type::{{closure}}
  16: rustc::traits::project::poly_project_and_unify_type
  17: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  18: rustc::traits::select::SelectionContext::evaluate_obligation
  19: <core::iter::Filter<I, P> as core::iter::iterator::Iterator>::next
  20: rustc_typeck::check::method::probe::ProbeContext::consider_candidates
  21: rustc_typeck::check::method::probe::ProbeContext::pick_method
  22: rustc_typeck::check::method::probe::ProbeContext::pick_core
  23: rustc_typeck::check::method::probe::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::probe_for_name
  24: rustc_typeck::check::FnCtxt::check_method_call
  25: rustc_typeck::check::FnCtxt::check_expr_kind
  26: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_lvalue_pref
  27: rustc_typeck::check::FnCtxt::check_expr_kind
  28: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_lvalue_pref
  29: rustc_typeck::check::FnCtxt::check_block_with_expected::{{closure}}
  30: rustc_typeck::check::FnCtxt::check_block_with_expected
  31: rustc_typeck::check::FnCtxt::check_expr_kind
  32: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_lvalue_pref
  33: rustc_typeck::check::FnCtxt::check_return_expr
  34: rustc_typeck::check::check_fn
  35: rustc_typeck::check::typeck_tables_of::{{closure}}
  36: rustc_typeck::check::typeck_tables_of
  37: rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::try_get
  38: rustc::ty::maps::TyCtxtAt::typeck_tables_of
  39: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::typeck_tables_of
  40: rustc_typeck::check::typeck_item_bodies
  41: rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_item_bodies<'tcx>>::try_get
  42: rustc::ty::maps::TyCtxtAt::typeck_item_bodies
  43: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::typeck_item_bodies
  44: rustc_typeck::check_crate
  45: rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}
  46: rustc_driver::driver::phase_3_run_analysis_passes
  47: rustc_driver::driver::compile_input
  48: rustc_driver::run_compiler
```

### LINKED ISSUE #44405: ICE when calling mutation method on map value indexed with mutable reference
When calling a mutation method on an value obtained from a HashMap with a mutable reference index, an ICE occurs.

Code:
```rust
use std::collections::HashMap;

fn main() {
    let mut map: HashMap<i32, Vec<i32>> = HashMap::new();
    let mut i = 1;
    map.insert(i, Vec::new());

    map[&mut i].push(3); // ICE occurs here
}
```
Compiler output:
```
error: internal compiler error: re-trying op failed
 --> src/main.rs:8:5
  |
8 |     map[&mut i].push(3);
  |     ^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.
```

Backtrace:
```
thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:437:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:380
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:390
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
   5: std::panicking::begin_panic_new
   6: rustc_errors::Handler::abort_if_errors
   7: rustc_passes::consts::check_crate
   8: rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}
   9: rustc_driver::driver::phase_3_run_analysis_passes
  10: rustc_driver::driver::compile_input
  11: rustc_driver::run_compiler
```

I believe that the correct behavior here would be for the compiler to report an error to the user on the indicated line due to IndexMut not being implemented for HashMap, which is what occurs when replacing `&mut i` with just `&i`.

rustc version:
```
rustc 1.20.0 (f3d6973f4 2017-08-27)
binary: rustc
commit-hash: f3d6973f41a7d1fb83029c9c0ceaf0f5d4fd7208
commit-date: 2017-08-27
host: x86_64-unknown-linux-gnu
release: 1.20.0
LLVM version: 4.0
```

### LINKED ISSUE #55265: ICE librustc/ty/subst.rs:462: Type parameter `DST/#0` (DST/0) out of range when substituting (root type=Some(DST)) substs=[ReErased, T, DST]
```rust
struct VTable<DST: ?Sized> {
 _to_dst_ptr: fn(*mut ()) -> *mut DST,
}

trait HasVTableFor<'a, DST: ?Sized 'a> {
 const VTABLE: &'a VTable<DST>;
}

impl<'a, T, DST: ?Sized 'a> HasVTableFor<'a, DST> for T {
 const VTABLE: &'a VTable<DST> = &VTable {
 _to_dst_ptr: |_: *mut ()| unsafe { std::mem::zeroed() },
 };
}

pub fn push<DST: ?Sized, T>() {
 <T as HasVTableFor<DST>>::VTABLE;
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=46b6ab803ff374f15c4f59e678d1a93d))

Errors:

```rust
 Compiling playground v0.0.1 (/playground)
warning: path statement with no effect
 --> src/lib.rs:16:5
 |
16 | <T as HasVTableFor<DST>>::VTABLE;
 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 |
 = note: #[warn(path_statements)] on by default

error: internal compiler error: librustc/ty/subst.rs:462: Type parameter `DST/#0` (DST/0) out of range when substituting (root type=Some(DST)) substs=[ReErased, T, DST]

thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:538:9
stack backtrace:
 0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
 at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
 1: std::sys_common::backtrace::print
 at libstd/sys_common/backtrace.rs:71
 at libstd/sys_common/backtrace.rs:59
 2: std::panicking::default_hook::{{closure}}
 at libstd/panicking.rs:211
 3: std::panicking::default_hook
 at libstd/panicking.rs:227
 4: rustc::util::common::panic_hook
 5: std::panicking::rust_panic_with_hook
 at libstd/panicking.rs:480
 6: std::panicking::begin_panic
 7: rustc_errors::Handler::span_bug
 8: rustc::util::bug::opt_span_bug_fmt::{{closure}}
 9: rustc::ty::context::tls::with_opt::{{closure}}
 10: rustc::ty::context::tls::with_context_opt
 11: rustc::ty::context::tls::with_opt
 12: rustc::util::bug::opt_span_bug_fmt
 13: rustc::util::bug::span_bug_fmt
 14: <rustc::ty::subst::SubstFolder<'a, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_ty
 15: <smallvec::SmallVec<A> as core::iter::traits::FromIterator<<A as smallvec::Array>::Item>>::from_iter
 16: rustc::ty::fold::TypeFoldable::fold_with
 17: rustc_mir::interpret::cast::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::cast
 18: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::run
 19: rustc_mir::const_eval::eval_body_using_ecx
 20: rustc_mir::const_eval::const_eval_provider
 21: rustc::ty::query::__query_compute::const_eval
 22: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval<'tcx>>::compute
 23: rustc::dep_graph::graph::DepGraph::with_task_impl
 24: rustc::ty::context::tls::with_related_context
 25: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
 26: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
 27: rustc::ty::query::TyCtxtAt::const_eval
 28: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::global_to_op
 29: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::const_value_to_op
 30: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::const_to_op
 31: rustc_mir::transform::const_prop::ConstPropagator::eval_constant
 32: rustc_mir::transform::const_prop::ConstPropagator::const_prop
 33: <rustc_mir::transform::const_prop::ConstPropagator<'b, 'a, 'tcx> as rustc::mir::visit::Visitor<'tcx>>::visit_statement
 34: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
 35: rustc_mir::transform::optimized_mir::{{closure}}
 36: rustc_mir::transform::optimized_mir
 37: rustc::ty::query::__query_compute::optimized_mir
 38: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::optimized_mir<'tcx>>::compute
 39: rustc::dep_graph::graph::DepGraph::with_task_impl
 40: rustc::ty::context::tls::with_related_context
 41: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
 42: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::try_get_query
 43: rustc::ty::query::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::optimized_mir
 44: rustc_metadata::encoder::<impl rustc_metadata::isolated_encoder::IsolatedEncoder<'a, 'b, 'tcx>>::encode_optimized_mir
 45: rustc_metadata::encoder::<impl rustc_metadata::isolated_encoder::IsolatedEncoder<'a, 'b, 'tcx>>::encode_info_for_item
 46: rustc::ty::context::tls::with_context
 47: <rustc_metadata::encoder::EncodeVisitor<'a, 'b, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
 48: rustc::hir::Crate::visit_all_item_likes
 49: rustc_metadata::encoder::encode_metadata
 50: rustc_metadata::cstore_impl::<impl rustc::middle::cstore::CrateStore for rustc_metadata::cstore::CStore>::encode_metadata
 51: rustc::ty::context::TyCtxt::encode_metadata
 52: rustc_codegen_llvm::base::write_metadata
 53: rustc::util::common::time
 54: rustc_codegen_llvm::base::codegen_crate
 55: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
 56: rustc::util::common::time
 57: rustc_driver::driver::phase_4_codegen
 58: rustc_driver::driver::compile_input::{{closure}}
 59: rustc::ty::context::tls::enter_context
 60: <std::thread::local::LocalKey<T>>::with
 61: rustc::ty::context::TyCtxt::create_and_enter
 62: rustc_driver::driver::compile_input
 63: rustc_driver::run_compiler_with_pool
 64: rustc_driver::driver::spawn_thread_pool
 65: rustc_driver::run_compiler
 66: <scoped_tls::ScopedKey<T>>::set
 67: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 68: __rust_maybe_catch_panic
 at libpanic_unwind/lib.rs:102
 69: rustc_driver::run
 70: rustc_driver::main
 71: std::rt::lang_start::{{closure}}
 72: std::panicking::try::do_call
 at libstd/rt.rs:59
 at libstd/panicking.rs:310
 73: __rust_maybe_catch_panic
 at libpanic_unwind/lib.rs:102
 74: std::rt::lang_start_internal
 at libstd/panicking.rs:289
 at libstd/panic.rs:392
 at libstd/rt.rs:58
 75: main
 76: __libc_start_main
 77: <unknown>
query stack during panic:
#0 [const_eval] const-evaluating `<T as HasVTableFor<'a, DST>>::VTABLE`
 --> src/lib.rs:16:5
 |
16 | <T as HasVTableFor<DST>>::VTABLE;
 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#1 [optimized_mir] processing `push`
end of query stack
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.31.0-nightly (424a749a0 2018-10-21) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by JohnTitor**
Closes #43623
Closes #44405

r? @Centril

### CODE REVIEW

File: src/test/ui/traits/trait-with-dst-2.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+struct VTable<DST: ?Sized> {
```
Comment (Centril) [inline src/test/ui/traits/trait-with-dst-2.rs]: I cannot get the ICE to reproduce with godbolt... I tried several versions but no luck.

Comment (JohnTitor) [reply src/test/ui/traits/trait-with-dst-2.rs]: Ah yes, it seems this doesn't cause an ICE (I tested with `nightly-2018-10-23` on my local). Should we remove this test and just close the issue?

File: src/test/ui/traits/trait-with-dst-2.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    _to_dst_ptr: fn(*mut ()) -> *mut DST,
+}
+
+trait HasVTableFor<'a, DST: ?Sized 'a> {
```
Comment (Centril) [inline src/test/ui/traits/trait-with-dst-2.rs]: Can you check whether fixing the missing comma here causes an ICE or not on 2018-10-13?
If not then let's remove the test and close the issue.

Comment (JohnTitor) [reply src/test/ui/traits/trait-with-dst-2.rs]: Hmm, sorry but what's the meaning of missing comma? Do you mean `?Sized + 'a`?

Comment (Centril) [reply src/test/ui/traits/trait-with-dst-2.rs]: No I meant `<'a, DST: ?Sized, 'a>` but `+` is probably what was intended. :)

Comment (JohnTitor) [reply src/test/ui/traits/trait-with-dst-2.rs]: Okay, I tested with `+` and comma but can't reproduce an ICE. (`nightly-2018-10-22` ~ `nightly-2018-10-24`)

Comment (Centril) [reply src/test/ui/traits/trait-with-dst-2.rs]: Alright; let's remove and close then.

**Pull request merged**